File: adapter/src/main/java/com/ncl/adapter/GenericAdapter.java
package com.ncl.adapter;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import com.ncl.adapter.exception.CellViewModelCreateException;
import java.util.Collection;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;


public class GenericAdapter<T extends CellViewModel> extends RecyclerView.Adapter<GenericViewHolder> {

    private final CellTypeManager<T> cellTypeManager;
    private LayoutInflater layoutInflater;
    private AdapteeCollection<T> collection;

    public GenericAdapter(CellTypeManager<T> cellTypeManager) {
        this(cellTypeManager, new ListAdapteeCollection<T>());
    }

    public GenericAdapter(CellTypeManager<T> cellTypeManager, AdapteeCollection<T> collection) {
        this.cellTypeManager = cellTypeManager;
        this.collection = collection;
    }

    @Override
    public int getItemCount() {
        return collection.size();
    }

    @Override
    public long getItemId(int position) {
        // TODO(Pablo): Do not return position return the real itemId if the content has.
        return position;
    }

    /**
     * Indicate to the RecyclerView the type of CellViewModel used to one position using a numeric value.
     *
     * @param position to analyze.
     * @return the id associated to the CellViewModel used to render the content given a position.
     */
    @Override
    public int getItemViewType(int position) {
        return cellTypeManager.getItemViewType(getItem(position));
    }

    /**
     * One of the two main methods in this class. Creates a GenericViewHolder instance with a
     * CellViewModel inside ready to be used. The CellTypeManager to create a GenericViewHolder using the
     * information given as parameter.
     *
     * @param viewGroup used to create the ViewHolder.
     * @param viewType  associated to the renderer.
     * @return ViewHolder extension with the CellViewModel it has to use inside.
     */
    @Override
    public GenericViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {

        if (layoutInflater == null) {
            // Reuse the inflater since it is associated to the same Activity Context
            layoutInflater = LayoutInflater.from(viewGroup.getContext());
        }
/*
        CellViewModel cellViewModel = cellTypeManager.buildGenericViewHolder(viewType);

        if (cellViewModel == null) {
            throw new CellViewModelCreateException("CellTypeManager have to return a not null viewHolder");
        }

        cellViewModel.onCreate(layoutInflater, viewGroup);
*/

        return cellTypeManager.buildGenericViewHolder(viewType, layoutInflater, viewGroup);
    }

    /**
     * Given a GenericViewHolder passed as argument and a position renders the view using the
     * CellViewModel previously stored into the GenericViewHolder.
     *
     * @param viewHolder with a CellViewModel class inside.
     * @param position   to render.
     */
    @SuppressWarnings("unchecked")
    @Override
    public void onBindViewHolder(@NonNull GenericViewHolder viewHolder, int position) {
        T cellViewModel = getItem(position);

        cellViewModel.onBound();
    }

    public T getItem(int position) {
        return collection.get(position);
    }

    /**
     * Add an element to the AdapteeCollection.
     *
     * @param element to add.
     * @return if the element has been added.
     */
    public boolean add(T element) {
        return collection.add(element);
    }

    /**
     * Remove an element from the AdapteeCollection.
     *
     * @param element to remove.
     * @return if the element has been removed.
     */
    public boolean remove(Object element) {
        return collection.remove(element);
    }

    /**
     * Add a Collection of elements to the AdapteeCollection.
     *
     * @param elements to add.
     * @return if the elements have been added.
     */
    public boolean addAll(Collection<? extends T> elements) {
        return collection.addAll(elements);
    }

    /**
     * Remove a Collection of elements to the AdapteeCollection.
     *
     * @param elements to remove.
     * @return if the elements have been removed.
     */
    public boolean removeAll(Collection<?> elements) {
        return collection.removeAll(elements);
    }

    /**
     * Remove all elements inside the AdapteeCollection.
     */
    public void clear() {
        collection.clear();
    }

    /**
     * Allows the client code to access the AdapteeCollection from subtypes of GenericAdapter.
     *
     * @return collection used in the adapter as the adaptee class.
     */
    protected AdapteeCollection<T> getCollection() {
        return collection;
    }

    public void setCollection(AdapteeCollection<T> collection) {
        if (collection == null) {
            throw new IllegalArgumentException("The AdapteeCollection configured can't be null");
        }

        this.collection = collection;
    }

    @Override
    public void onViewRecycled(GenericViewHolder holder) {
        super.onViewRecycled(holder);
        //holder.getCellViewModel().onRecycle();
    }

}

File: adapter/src/main/java/com/ncl/adapter/GenericAdapter.kt
package com.ncl.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import java.util.HashMap
import androidx.annotation.NonNull
import androidx.recyclerview.widget.RecyclerView


class GenericAdapter<VM : CellViewModel<GenericViewHolder>> : RecyclerView.Adapter<GenericViewHolder>() {

    private var collection: ArrayList<VM>
    private var layoutInflater: LayoutInflater? = null
    private val classCellTypeMap: HashMap<Class<out CellViewModel<GenericViewHolder>>, Int>
    private val typeInstanceMap: HashMap<Int, VM>

    init {
        this.collection = ArrayList()
        this.classCellTypeMap = HashMap()
        this.typeInstanceMap = HashMap()
    }

    override fun getItemCount(): Int {
        return collection.size
    }

    override fun getItemId(position: Int): Long {
        // TODO(Pablo): Do not return position return the real itemId if the content has.
        return position.toLong()
    }

    fun getItem(position: Int): VM {
        return collection.get(position)
    }

    override fun getItemViewType(position: Int): Int {
        val cellViewModel = getItem(position)
        val cellViewModelClass = cellViewModel.javaClass
        var adapterViewType: Int? = classCellTypeMap[cellViewModelClass]

        if (adapterViewType == null) {
            adapterViewType = classCellTypeMap.size
            // register a cell viewType for this model class
            classCellTypeMap[cellViewModelClass] = adapterViewType
            typeInstanceMap[adapterViewType] = cellViewModel
        }

        return adapterViewType
    }

    override fun onCreateViewHolder(@NonNull viewGroup: ViewGroup, viewType: Int): GenericViewHolder {

        val inflater: LayoutInflater = layoutInflater ?: run {
            val newInflater = LayoutInflater.from(viewGroup.context)
            layoutInflater = newInflater
            newInflater
        }

        // TODO(Pablo): Remove the force unwrapping and return a Default Empty ViewHolder
        return typeInstanceMap[viewType]!!.createViewHolder(inflater, viewGroup)
    }

    override fun onBindViewHolder(@NonNull viewHolder: GenericViewHolder, position: Int) {
        val cellViewModel = getItem(position)
        viewHolder.cellViewModel = cellViewModel
        cellViewModel.bindViewHolder(this, viewHolder)
    }

    override fun onViewRecycled(holder: GenericViewHolder) {
        super.onViewRecycled(holder)
        holder.cellViewModel?.unbindViewHolder(this, holder)
    }


    // region: Collection Methods

    /**
     * Add an element to the AdapteeCollection.
     *
     * @param element to add.
     * @return if the element has been added.
     */
    fun add(element: VM): Boolean {
        return collection.add(element)
    }

    /**
     * Remove an element from the AdapteeCollection.
     *
     * @param element to remove.
     * @return if the element has been removed.
     */
    fun remove(element: Any): Boolean {
        return collection.remove(element)
    }

    /**
     * Add a Collection of elements to the AdapteeCollection.
     *
     * @param elements to add.
     * @return if the elements have been added.
     */
    fun addAll(elements: Collection<VM>): Boolean {
        return collection.addAll(elements)
    }

    /**
     * Remove a Collection of elements to the AdapteeCollection.
     *
     * @param elements to remove.
     * @return if the elements have been removed.
     */
    fun removeAll(elements: Collection<*>): Boolean {
        return collection.removeAll(elements)
    }

    /**
     * Remove all elements inside the AdapteeCollection.
     */
    fun clear() {
        collection.clear()
    }

    /**
     * Allows the client code to access the AdapteeCollection from subtypes of GenericAdapter.
     *
     * @return collection used in the adapter as the adaptee class.
     */
    protected fun getCollection(): ArrayList<VM> {
        return collection
    }

    fun setCollection(collection: ArrayList<VM>) {
        this.collection = collection
    }

    // endregion

}
